### PR TITLE
SI-7292 Deprecate octal escape literals

### PR DESCRIPTION
**Pull request by soc**

### CODE REVIEW

**Comment by soc** [comment]
Thread on the mailing list: https://groups.google.com/forum/#!topic/scala-internals/bxmRdZFpFDo
The issues in the tests regarding `/` vs. `\` are also mentioned there.
@paulp, could you review it?

**Comment by som-snytt** [comment]
The normalization of slashes in test output was due to backslash in Windows paths.

The commit for Testicolor has a comment about a test breaking; apparently, I deferred the issue; but I will give it more thought.  In my other branch I had taken out "indisciminate [sic] replacement of backslash".

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala [hunk 1]
```diff
@@ -800,6 +800,17 @@ trait Scanners extends ScannersCommon {
...
+            if (i == '\n') {
+              "\\n"
+            } else {
+              val hexStr = i.toHexString
+              val padding = "0" * (4 - hexStr.length)
+              s"\\u$padding$hexStr"
+            }
+          }
+          if (settings.future.value) syntaxError(charOffset - 3, s"Octal escape literals are not supported anymore, use ${int2unicodeEscape(oct)} instead.")
+          else deprecationWarning(charOffset - 3, s"Octal escape literals are deprecated, use ${int2unicodeEscape(oct)} instead.")
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: The only good duplication is a dead duplication. You can deduce what "start" is.

```
val alt = if (oct == LF) LF else "\\u%04x" format oct
def msg(what: String) = s"Octal escape literals are $what, use $alt instead."
if (settings.future.value)
  syntaxError(start, msg("unsupported"))
else
  deprecationWarning(start, msg("deprecated"))
```

**Comment by soc** [comment]
Thanks, good points. I'll fix them in a few hours when I'm done with the Avian stuff.

**Comment by soc** [comment]
@paulp, @som-snytt, I updated and pushed the changes.

File: src/compiler/scala/tools/nsc/ast/parser/Scanners.scala [hunk 2]
```diff
@@ -800,6 +800,13 @@ trait Scanners extends ScannersCommon {
               nextChar()
             }
           }
+          val alt = if (oct == LF) "\\n" else "\\u%04x" format oct
+          val start = charOffset - 3
```
Comment (paulp) [inline src/compiler/scala/tools/nsc/ast/parser/Scanners.scala]: You did not correctly deduce what start was! Look at your checkfiles, where the carat is pointing into the middle of the numbers. The carat should always be pointing at the start of the invalid sequence.

**Comment by soc** [comment]
@paulp, @som-snytt, updated again. Sorry for the trouble.

**Comment by paulp** [comment]
LGTM now in terms of the code itself; the remaining question is whether we actually agreed to deprecate them. It wasn't clear to me that was the case, but I wasn't that plugged in. Any opinions? @adriaanm @retronym?

**Comment by adriaanm** [comment]
Dare I point to https://github.com/scala/scala-dist/pull/20?

**Comment by soc** [comment]
@paulp I didn't receive much (any?) feedback, that's when I decided to move forward with it.
@adriaanm  It is slightly related to the octal literals which have already been deprecated in the sense that they are octal too, but the octal escape literals are even more obscure.
In terms of the spec: remove "A character with Unicode between 0 and 255 may also be represented by an octal escape, i.e. a backslash ‘’ followed by a sequence of up to three octal characters." from "1.3.6 Escape Sequences" and " or the octal escape '\12'" from "1.3.4 Character Literals".

**Comment by adriaanm** [comment]
I talked to Martin about this and he feels the confusion of straying further from Java on this is more expensive than the gain of cleaning this up in Scala. What's the motivation to deprecate this? Does the notation cause problems? Is it a source of bugs?

I'm closing this for now. Please reopen with counter-arguments.

**Comment by soc** [comment]
@adriaanm, these are my counter-arguments, but I lack the privilege to reopen this PR, could you do that for me?
- We got rid of octal literals which are similarly confusing. I have never heard a _single_ person complaining (or even asking) about what happened with octal literals. It's reasonable to assume that it will be the same for an even more obscure feature.
- The parsing rules of octal escapes are unintuitive and misleading for unaware readers (e. g. it's not fixed-length like `\uXXXX`, but stops accepting symbols as soon as it encounters some char it doesn't like. Example: `\377` parses as one octal escape literal, `\378` parses as an octal escape and `8` (because 8 is not octal, duh), `400` parses as an octal escape and a `0` (`400` would be valid octal number, but–for whatever reasons–the literal is capped to the first 255 numbers)).
- Those who want it, can implement it via string interpolation
- "straying from Java" is reality already: See Scala's change to floating point literals, Scala's removal of octal literals, Java's addition of binary literals, Java's addition of underscores-in-literals.
- If you look at some of our "competitors", their literal syntax is even further removed from Java syntax.
- In general, we need less magic to happen between what-the-developer-writes and what-the-compiler-sees. See the famous let's-convert-unicode-escapes-in-comments puzzler for a similar case where we are trying too hard to be Java. (Which–again–no other languages even considers emulating.)

**Comment by som-snytt** [comment]
Hate to say it, but is it worth a language.feature barrier?

To me, the issue isn't "straying from Java" so much as "straying from our '70s geek roots."

I don't know if they made octal jokes on "That '70s Show".

Maybe certain features could be enabled if the compiler is invoked with `scalac -classic`.

I'm going to go google "Octalpussy" now and see what comes up. Well, I guess that's already '80s.

**Comment by adriaanm** [comment]
Ok, sounds convincing. Could you do a poll on the mailing lists? If no one speaks up, we can reopen and deprecate in 2.11.0-M4 as far as I'm concerned. /CC @odersky

**Comment by adriaanm** [comment]
(I mean the user lists, I saw https://groups.google.com/forum/#!topic/scala-internals/bxmRdZFpFDo/discussion)

**Comment by soc** [comment]
@som-snytt I don't think it is worth it. We didn't do it for octal literals and it went through without issues or complaints.

@adriaanm Alright, thanks! I'll do it in a few hours, it's 02:34 and I really need to catch some sleep.

**Comment by som-snytt** [comment]
> We didn't do it for octal literals

You didn't hear me cry, They got rid of octal literals!

Probably I won't need it for the posix permissions in nio.2.

But part of me wishes I _did_ need it.

**Comment by soc** [comment]
https://groups.google.com/forum/#!topic/scala-user/Y5oU6mIrFZI

**Comment by soc** [comment]
@adriaanm The mailing list seems to be in favor of removing it.

**Comment by adriaanm** [comment]
Sorry I didn't know reopening required special permission. I'll reopen once the release is out. Don't want to steal cycles from that this week.

**Comment by soc** [comment]
No problem, thanks!

**Comment by soc** [comment]
@adriaanm Scala 2.10.2 seems to be out, can you reopen this PR?

**Comment by JamesIry** [comment]
Only mostly out, but I think we can reopen.

**Comment by soc** [comment]
@adriaanm, @som-snytt, @JamesIry Is there anything missing to move this forward?

**Comment by SethTisue** [comment]
so happy this and similar cleanups/removals are going through

**Pull request merged**